lib/TPP/Transforms/LowerPacksAndUnpacksWithoutTranspose.cpp

```diff
@@ -0,0 +1,173 @@
+#include "TPP/Transforms/Utils/ValueUtils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tensor/Utils/Utils.h"
+#include "mlir/Dialect/Utils/IndexingUtils.h"
+#include "mlir/IR/AffineExpr.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/Interfaces/DestinationStyleOpInterface.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Casting.h"
+#include <cstdint>
+using namespace mlir;
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_LOWERPACKSANDUNPACKSWITHOUTTRANSPOSE
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+namespace {
+
+/// Wrapper around linalg::lowerPack which undoes the transpose that might have
+/// happened. Single user genericOp's indexing_maps is corrected accordingly.
+void lowerPackWithoutTranspose(tensor::PackOp packOp,
+                               linalg::GenericOp genericOp, uint operandIdx,
+                               PatternRewriter &rewriter) {
+  auto packInversionPerm = tensor::getPackInverseDestPerm(packOp);
+
+  auto res = linalg::lowerPack(rewriter, packOp);
+
+  if (res->transposeOp) {
+    // Forget about the permutation of the dims on expandShapeOp.
+    rewriter.replaceAllOpUsesWith(res->transposeOp, res->expandShapeOp);
+
+    // Invert corresponding transposed accesses by the single-user, genericOp.
+    auto indexingMaps = genericOp.getIndexingMapsArray();
+    auto packInverseMap =
+        AffineMap::getPermutationMap(packInversionPerm, rewriter.getContext());
+    auto normalizedIndexingMap =
+        packInverseMap.compose(indexingMaps[operandIdx]);
+
+    SmallVector<Attribute> maps = llvm::to_vector(genericOp.getIndexingMaps());
+    maps[operandIdx] = AffineMapAttr::get(normalizedIndexingMap);
+    genericOp.setIndexingMapsAttr(ArrayAttr::get(rewriter.getContext(), maps));
+  }
+}
+
+struct LowerPackOnInputsWithoutTranspose
+    : public OpRewritePattern<linalg::GenericOp> {
+  // Is only called with single-user packOp operands, so callback can always
+  // find the (use by the) linalg.generic that is the target of the pattern.
+  using ControlFn = std::function<bool(tensor::PackOp)>;
+  ControlFn controlFn;
+
+  LowerPackOnInputsWithoutTranspose(MLIRContext *context,
+                                    ControlFn controlFn = nullptr,
+                                    PatternBenefit benefit = 1)
+      : OpRewritePattern(context, benefit), controlFn(std::move(controlFn)) {}
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    bool modifiedAnOperand = false;
+    for (auto operandIdx : {0, 1}) {
```

Comment:
That's an oddly specific pattern. I think the pattern should be either more precise or more generic.

Reply:
This is based on your suggestion to have separate patterns.

Reply:
I mean the explicit operands 0 and 1

Reply:
I see - the current thing is incorrect in case of non-binary input generics. Before I had the guard that this only applied to ContractionInterface ops but I removed it as I thought it was not necessary.

Will fix this so that it iterates over all inputs of the generic (and correspondingly fix the output/unpack pattern).

Reply:
Made both patterns agnostic to the arity. If you could check if this looks right, that would be appreciated.